packages/network/src/helpers/Metric.ts

```diff
@@ -0,0 +1,234 @@
+import EventEmitter from 'eventemitter3'
+import { set } from 'lodash'
+import { scheduleAtFixedRate } from './scheduler'
+
+export abstract class Sampler {
+
+    protected metric: Metric
+    private listener: any
+
+    constructor(metric: Metric) {
+        this.metric = metric
+        this.listener = (value: number) => this.onRecord(value)
+    }
+
+    start(_now: number): void {
+        this.metric.on('record', this.listener)
+    }
+
+    stop(_now: number): void {
+        this.metric.off('record', this.listener)
+    }
+
+    protected abstract onRecord(value: number): void
+
+    abstract getAggregatedValue(): number | undefined
+}
+
+export type MetricsDefinition = Record<string,Metric>
+
+interface MetricEvents {
+    record: (value: number) => void
+}
+
+export class Metric {
+
+    private latestValue: number | undefined
+    private eventEmitter: EventEmitter<MetricEvents> = new EventEmitter()
+    private samplerFactory: (metric: Metric) => Sampler
+
+    constructor(samplerFactory: (metric: Metric) => Sampler, initialValue?: number) {
+        this.samplerFactory = samplerFactory
+        this.latestValue = initialValue
+    }
+
+    record(value: number): void {
+        this.latestValue = value
+        this.eventEmitter.emit('record', value)
+    }
+
+    getLatestValue(): number | undefined {
+        return this.latestValue
+    }
+
+    on<T extends keyof MetricEvents>(eventName: T, listener: MetricEvents[T]): void {
+        this.eventEmitter.on(eventName, listener as any)
+    }
+
+    off<T extends keyof MetricEvents>(eventName: T, listener: MetricEvents[T]): void {
+        this.eventEmitter.off(eventName, listener as any)
+    }
+
+    createSampler(): Sampler {
+        return this.samplerFactory(this)
+    }
+}
+
+/*
+ * Sum of all records within a sampling period. 
+ *
+ * E.g. count of failed connections
+ */
+class CountSampler extends Sampler {
+
+    protected sum: number = 0
+
+    protected onRecord(value: number): void {
+        this.sum += value
+    }
+    
+    getAggregatedValue(): number | undefined {
+        return this.sum
+    }
+}
+
+export class CountMetric extends Metric {
+    constructor() {
+        super((m) => new CountSampler(m))
+    }
+}
+
+/*
+ * Average of all records within a sampling period. If the sampling data is
+ * continuous, you may want to use LevelSampler instead.
+ *
+ * E.g. average latency
+ */
+class AverageSampler extends CountSampler {
```

Comment:
I'd prefer it to keep track of `sum` itself and not inherit. Maybe just simpler from inheritance analysis viewpoint.
```suggestion
class AverageSampler extends CountSampler {
```